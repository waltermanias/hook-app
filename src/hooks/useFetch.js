import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });

    fetch(url)
      .then((response) => response.json())
      .then((data) => {
        if (isMounted.current) {
          setState({
            data,
            loading: false,
            error: null,
          });
        }
      })
      .catch((err) => {
        if (isMounted.current) {
          setState({
            data: null,
            loading: false,
            error: err,
          });
        }
      });
  }, [url]);

  return state;
};

// https://www.breakingbadapi.com/api/quotes/1
