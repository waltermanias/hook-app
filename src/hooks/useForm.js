import { useState } from "react";

export const useForm = (initialState = {}) => {
  const [formValues, setForm] = useState(initialState);

  const reset = () => {
    setForm(initialState);
  };

  const handleInputChange = (e) => {
    setForm({
      ...formValues,
      [e.target.name]: e.target.value,
    });
  };

  return {
    formValues,
    handleInputChange,
    reset,
  };
};
