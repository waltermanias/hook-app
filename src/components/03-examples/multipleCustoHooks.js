import React from "react";
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";
import "./hooks.css";

export const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);

  const { error, data, loading } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { author, quote } = !!data && data[0];

  return (
    <>
      <h1>Breaking Bad Quotes</h1>
      <hr />

      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <blockquote className="blockquote text-right">
          <p className="mb-2">{quote}</p>
          <footer className="blockquote-footer">{author}</footer>
        </blockquote>
      )}

      <button className="btn btn-primary" onClick={increment}>
        Next Quote
      </button>
    </>
  );
};
