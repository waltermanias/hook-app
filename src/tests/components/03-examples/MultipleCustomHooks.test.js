import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/multipleCustoHooks";
import { useFetch } from "../../../hooks/useFetch";

jest.mock("../../../hooks/useFetch");

describe("MultipleCustomHooks", () => {
  it("should render the component", () => {
    useFetch.mockReturnValue({ data: null, loading: true, error: null });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should show the info", () => {
    useFetch.mockReturnValue({
      error: null,
      data: [{ author: "test-author", quote: "test-quote" }],
      loading: false,
    });

    const wrapper = shallow(<MultipleCustomHooks />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('.mb-2').text().trim()).toBe("test-quote");
    expect(wrapper.find('footer').text().trim()).toBe("test-author");
  });


});
