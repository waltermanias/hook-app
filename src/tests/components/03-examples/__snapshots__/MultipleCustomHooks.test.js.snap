// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleCustomHooks should render the component 1`] = `
<Fragment>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</Fragment>
`;

exports[`MultipleCustomHooks should show the info 1`] = `
<Fragment>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <blockquote
    className="blockquote text-right"
  >
    <p
      className="mb-2"
    >
      test-quote
    </p>
    <footer
      className="blockquote-footer"
    >
      test-author
    </footer>
  </blockquote>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</Fragment>
`;
